ocaml/typing/typemode.ml

```diff
@@ -40,23 +41,24 @@ let transl_mode_annots modes =
   loop Alloc.Const.Option.none modes.txt
 
 let transl_global_flags gfs =
-  let rec loop (acc : Global_flag.t) : _ -> Global_flag.t = function
+  let rec loop (acc : Global_flag.t) = function
     | [] -> acc
     | m :: rest ->
       let { txt; loc } = (m : Mode_expr.Const.t :> _ Location.loc) in
-      let acc =
+      let acc : Global_flag.t =
         match txt with
         | "global" -> (
           Jane_syntax_parsing.assert_extension_enabled ~loc Mode ();
           match acc with
-          | Unrestricted -> Global_flag.Global
+          | Unrestricted -> Global
+          (* Duplicated modality is not an error, just silly and thus a warning.
+             As we introduce more modalities, it might be in general difficult
+             to detect all redundant modalities, but we should do our best. *)
```

Comment:
I don't understand why duplicated mode is an error but duplicated modality is a warning. I find consistency to be compelling here. Is there something I'm not considering?

Reply:
I understand `local local` as assembling axes: Both `local` specify something in the locality axis, and you can't specify the locality axis twice. Sure, we can specially treat the case where the two specifications match, but I doubt it's worth the extra effort.

Modalities on the other hand are mappings. `global global` means `global (global (m))` where `m` is the mode of the record. It's silly but still sensible.

Reply:
I don't think we should allow duplicate modalities until there exists a modality for which it is meaningful, and even then we should only allow it for the meaningful modalities.

Reply:
Some modalities are useful to duplicate. For example, the `regional` modality would be useful under https://github.com/ocaml-flambda/flambda-backend/pull/1815 : if a record is at region `n`, a field with `regional regional` would be at region `n-2`. 

The following is me thinking out loud, that @lpw25 might have opinions on:

Now, if a user writes `type r = {x : string @@ global regional regional}`, that gives field's mode `regional (regional (global (m)))`, where `m` is the record's mode. That should be equivalent to `global (m)`.

If a user writes `type r = {x  : string @@ regional regional global}`, that gives field's mode `global (regional (regional (m)))`, which is still equivalent to `global (m)`. 

Well, I guess my gist is that, modalities are mappings and could have rich behaviours. As such, redundant modality detection is likely to be ad-hoc and incomplete. Therefore, it's better to issue warnings instead of errors. Also, those programs are not "wrong" anyway.

Reply:
Well, on a second thought, maybe we should restrict user to specify only one modality per axis? That should be easy to do, and not too restrictive. The aforementioned `regional regional` will be replaced by `regional(2)`.

If we agree to this restriction, I will be OK with taking violation of it as errors.

Reply:
I agree that there are modalities that are not idempotent and pairs of modalities that do not commute. Possibly we should avoid supporting such modalities of course, because they are difficult to think about for the user. I agree that warnings for redundent modalities is the way to go -- although I think we don't really need that in the MVP.

Whether redundent modes should be a warning or an error depends on the details of our full mode expression language. In some approaches it is essentially a kinding error to repeat a mode, in other approaches it is merely redundant. I'm leaning towards an approach that makes it a kinding error, but I also think we can just do nothing for the MVP and address this when we add mode polymorphism.

Reply:
For the record, I asked around, and all warnings are enabled by default in our internal build, unless specified otherwise. So this should be fine.